certbot-apache/certbot_apache/override_centos.py

```diff
@@ -47,6 +53,71 @@ def get_parser(self):
             self.aug, self.option("server_root"), self.option("vhost_root"),
             self.version, configurator=self)
 
+    def _deploy_cert(self, *args, **kwargs):
+        """
+        Override _deploy_cert in order to ensure that the Apache configuration
+        has "LoadModule ssl_module..." before parsing the VirtualHost configuration
+        that was created by Certbot
+        """
+        super(CentOSConfigurator, self)._deploy_cert(*args, **kwargs)
+        if self.version < (2, 4, 0):
+            self._deploy_loadmodule_ssl_if_needed()
+
+
+    def _deploy_loadmodule_ssl_if_needed(self):
+        """
+        Add "LoadModule ssl_module <pre-existing path>" to main httpd.conf if
+        it doesn't exist there already.
+        """
+
+        loadmods = self.parser.find_dir("LoadModule", "ssl_module", exclude=False)
+
+        # We should have at least one LoadModule ssl_module in the config
+        loadmod_args = []  # type: List[str]
+        loadmod_path = None
+        for m in loadmods:
+            noarg_path = m.rpartition("/")[0]
+            path_args = self.parser.get_all_args(noarg_path)
+            if loadmod_args:
+                if loadmod_args != path_args:
+                    logger.info("Multiple different LoadModule directives for mod_ssl "
+                                "were found. If you encounter issues with resulting "
+                                "configuration, it's suggested to move the LoadModule "
+                                "ssl_module directive to the beginning of main Apache "
+                                "configuration file at /etc/httpd/conf/httpd.conf")
+                    return
+            else:
+                loadmod_args = path_args
+                loadmod_path = noarg_path
+            if self.parser.loc["default"] in noarg_path:
+                # LoadModule already in the main configuration file, NOOP
+                return
+
+        if not loadmod_args:
+            # Do not try to enable mod_ssl
+            return
+
+        rootconf_ifmod = self.parser.get_ifmod(
+            parser.get_aug_path(self.parser.loc["default"]),
+            "!mod_ssl.c", beginning=True)
+        # parser.get_ifmod returns a path postfixed with "/", remove that
+        self.parser.add_dir(rootconf_ifmod[:-1], "LoadModule", loadmod_args)
+        self.save_notes += "Added LoadModule ssl_module to main configuration.\n"
+
+        # Wrap LoadModule mod_ssl inside of <IfModule !mod_ssl.c> if it's not
+        # configured like this already.
+        if loadmod_path and "ifmodule" not in loadmod_path.lower():
+            sslconf_path = loadmod_path.split("/directive")[0]
+            # Remove the old LoadModule directive
+            self.aug.remove(loadmod_path)
+
+            # Create a new IfModule !mod_ssl.c
+            ssl_ifmod = self.parser.get_ifmod(sslconf_path, "!mod_ssl.c", beginning=True)
```

Comment:
This is certainly a corner case, but I encountered a minor issue with the code here. In `ssl.conf`, I had:
```
LoadModule ssl_module modules/mod_ssl.so

<IfModule !mod_ssl.c>
LoadModule ssl_module modules/mod_ssl.so
</IfModule>
```
When this code was run, it resulted in:
```
<IfModule !mod_ssl.c>
LoadModule ssl_module modules/mod_ssl.so
LoadModule ssl_module modules/mod_ssl.so
</IfModule>
```
Can we always add a new `IfModule` block specifically for the unwrapped `LoadModule `directive?

Reply:
I changed this in a way that paths that already have an existing `<IfModule !mod_ssl.c>` `LoadModule ssl_module ...` under them, don't get a new `LoadModule` added within. We basically care about having the `ssl_module` loaded in the main `httpd.conf`, and about not confusing the user too much.